libpod/image/pull.go

```diff
@@ -198,12 +200,62 @@ func (ir *Runtime) pullGoalFromImageReference(ctx context.Context, srcRef types.
 
 // pullImageFromHeuristicSource pulls an image based on inputName, which is heuristically parsed and may involve configured registries.
 // Use pullImageFromReference if the source is known precisely.
-func (ir *Runtime) pullImageFromHeuristicSource(ctx context.Context, inputName string, writer io.Writer, authfile, signaturePolicyPath string, signingOptions SigningOptions, dockerOptions *DockerRegistryOptions, label *string) ([]string, error) {
+func (ir *Runtime) pullImageFromHeuristicSource(ctx context.Context, config *PullConfig) ([]string, error) {
 	span, _ := opentracing.StartSpanFromContext(ctx, "pullImageFromHeuristicSource")
 	defer span.Finish()
+	sc := GetSystemContext(config.SignaturePolicyPath, config.Authfile, false)
+	goal, err := ir.getPullGoalFromHeuristicSource(ctx, sc, config.Name)
+	if err != nil {
+		return nil, err
+	}
+	return ir.doPullImage(ctx, sc, *goal, config.Writer, config.SigningOptions, config.DockerOptions, config.Label)
+}
 
+// pullImageIfOutOfDate checks an image is out of date with its source,
+// and if so, pulls an updated image.
+func (ir *Runtime) pullImageIfOutOfDate(ctx context.Context, localImg *storage.Image, config *PullConfig) ([]string, bool, error) {
+	span, _ := opentracing.StartSpanFromContext(ctx, "pullImageFromHeuristicSourceIfNecessary")
+	defer span.Finish()
+
+	sc := GetSystemContext(config.SignaturePolicyPath, config.Authfile, false)
+
+	goal, err := ir.getPullGoalFromHeuristicSource(ctx, sc, config.Name)
```

Comment:
BTW what about `goal.pullAllPairs`? That’s another feature that needs to be handled here, somehow.